### PR TITLE
Funct for get a Profile belonging to a User moved to db module and some style updates

### PR DESCRIPTION
**Pull request by luchoman08**
Changes summary:

- [Clippy warnings fixed](https://github.com/luchoman08/realworld-rust-rocket/commit/4ebec7d01c06c22619c59d01e46f7f4e1b68a7b6).
- Minnor style changes at [here](https://github.com/luchoman08/realworld-rust-rocket/commit/e35bfb41f7c21d56159794146b150f292220c88b) and [here](https://github.com/luchoman08/realworld-rust-rocket/commit/20207bda7d0c2f457de4e67e2b38c26a9322b471)
- Redeability changes [here](https://github.com/luchoman08/realworld-rust-rocket/commit/7a7d3ccb97babe053c291d6b8af2e078470c4b39) and [here](https://github.com/TatriX/realworld-rust-rocket/commit/e41de0100427cbb0b3cfe1311b98b91b0b008069)
-  Moved function for return a profile based in a user from models (method of User) to db module [here](https://github.com/TatriX/realworld-rust-rocket/commit/6f1d5e084544bc74c19445f740ff66a7360f0922)

Posdata: All test from [here](https://github.com/gothinkster/realworld/tree/master/api) are passed

### CODE REVIEW

**Comment by luchoman08** [comment]
This pull request have the changes of [this closed pull request](https://github.com/TatriX/realworld-rust-rocket/pull/11) 

Previous Changes:
- Function [get_comments](https://github.com/luchoman08/realworld-rust-rocket/blob/17ea029102161e2b0c25ca1688de5e0855d9942d/src/routes/articles.rs) now have `Optional<Auth>` for deduce de current user id and append the `following` option to author of comment 
- Article Json`s author and Comments`s author now are type of `Profile` instead of type `User` [see this commit](https://github.com/TatriX/realworld-rust-rocket/pull/12/commits/367080d07c39301ef0ddd34b0db59815b47f7073) and [this other](https://github.com/TatriX/realworld-rust-rocket/pull/12/commits/17ea029102161e2b0c25ca1688de5e0855d9942d)
- Tag list property at edit article is optional field, previously an error was returned if an article was eddited whitout tags property ([see this commit](https://github.com/TatriX/realworld-rust-rocket/pull/12/commits/edaa8cefbf374281105d9e8c3a08b23cdb8f3e82))

File: src/models/user.rs [hunk 1]
```diff
@@ -31,6 +30,21 @@ pub struct Profile {
     image: Option<String>,
     following: bool,
 }
+impl Profile {
```
Comment (TatriX) [inline src/models/user.rs:33]: What is the point of this constructor?

File: src/routes/articles.rs [hunk 2]
```diff
@@ -138,8 +138,9 @@ pub fn delete_comment(slug: String, id: i32, auth: Auth, conn: db::Conn) {
 }
 
 #[get("/articles/<slug>/comments")]
-pub fn get_comments(slug: String, conn: db::Conn) -> Json<JsonValue> {
-    let comments = db::comments::find_by_slug(&conn, &slug);
+pub fn get_comments(slug: String, conn: db::Conn, auth: Option<Auth>) -> Json<JsonValue> {
+    let auth_id = auth.and_then(|auth| Some(auth.id));
```
Comment (TatriX) [inline src/routes/articles.rs:142]:
```rust
     let auth_id = auth.map(|auth| auth.id); 
```
should work.

File: src/models/user.rs [hunk 3]
```diff
@@ -50,12 +64,11 @@ impl User {
             token,
         }
     }
-
-    pub fn to_profile(self, following: bool) -> Profile {
+    pub fn to_profile(&self, following: bool) -> Profile {
```
Comment (TatriX) [inline src/models/user.rs:67]: Could you please check if it's possible to use `into_profile` instead.
```rust
pub fn into_profile(self, following: bool) -> Profile { 
 Profile {
            username: self.username,
            bio: self.bio,
            image: self.image,
            following,
        }
}
```

File: src/db/users.rs [hunk 4]
```diff
@@ -1,11 +1,48 @@
-use crate::models::user::User;
+use crate::models::user::{Profile, User};
 use crate::schema::users;
 use crypto::scrypt::{scrypt_check, scrypt_simple, ScryptParams};
 use diesel::dsl::{exists, select};
 use diesel::pg::PgConnection;
 use diesel::prelude::*;
 use serde::Deserialize;
 
+/// Return a `Profile` for a given `user` adding the `following` property.
```
Comment (TatriX) [inline src/db/users.rs:9]: This comment looks like an overkill.
Also I'm not entirely sure what is the point of this function.
 It looks like what you want is something like `user.to_profile(is_following(conn, user, user_id))`

File: src/models/user.rs (hunk 3 above)
Comment (luchoman08) [reply src/models/user.rs:67]: When this method have `self` as first argument, this warning is displayed by [clippy](https://github.com/rust-lang/rust-clippy): [Imgur](https://i.imgur.com/0aahORt.png)

File: src/db/users.rs (hunk 4 above)
Comment (luchoman08) [reply src/db/users.rs:9]: You're right, I will eliminate that function

File: src/routes/articles.rs (hunk 2 above)
Comment (luchoman08) [reply src/routes/articles.rs:142]: Thanks, good tip

File: src/models/user.rs (hunk 3 above)
Comment (TatriX) [reply src/models/user.rs:67]: Yes, that's why I suggested using name `into_profile`, because by convention it consumes the receiver. Check this: https://rust-lang-nursery.github.io/api-guidelines/naming.html

Comment (luchoman08) [reply src/models/user.rs:67]: Thank you very much, it's like that

File: src/db/users.rs (hunk 4 above)
Comment (luchoman08) [reply src/db/users.rs:9]: I have been doing the refactoring, however I have observed that the function to_profile_for is useful, since the `author` field in `Comments` and the `Articles` is of type `Profile` (not `User`), the is_following function is used in association with the `into_profile` method at least six times, and in some cases the code becomes repetitive.

For example, in the case of `find_by_slug` at `db::comments` the code whitout `to_profile_for` would look like this:
```rust
pub fn find_by_slug(conn: &PgConnection, slug: &str, current_user_id: Option<i32>) -> Vec<CommentJson> {
    use crate::db::profiles::is_following;
    let result = comments::table
        .inner_join(articles::table)
        .inner_join(users::table)
        .select((comments::all_columns, users::all_columns))
        .filter(articles::slug.eq(slug))
        .get_results::<(Comment, User)>(conn)
        .expect("Cannot load comments");

    result
        .into_iter()
        .map(|(comment, author)| {
            let following =
            current_user_id.map_or(false, |user_id| is_following(conn, &author, user_id));
            comment.attach(author.into_profile(following))
        }
        )
        .collect()
}
```
Whit `to_profile_for` function this should looks like this
```rust
...
   result
        .into_iter()
        .map(|(comment, author)| comment.attach(to_profile_for(conn, &author, current_user_id))
        .collect()
...
```
Basically I'm relying on the function [toProfileJSONFor](https://github.com/gothinkster/node-express-realworld-example-app/blob/master/models/User.js) of the implementation of the nodeJs backend.

Comment (luchoman08) [reply src/db/users.rs:9]: Also, as shown in this  [Imgur](https://i.imgur.com/Z3Rwpgg.png)  image, the populate function also needs to be modified to receive an optional argument from the currently loggedin user  and correctly add the `comment.author.following` property, otherwise, even if the user is followed, the option to follow will always be shown, "unfollow" option will never be shown. 
I do not know if I'm foolishly insisting on the function `into_profile_for`, I hope to be instructed if there is another option, or the use of the functions `is_following` and `into_profile` together is enough.

File: src/models/user.rs (hunk 1 above)
Comment (luchoman08) [reply src/models/user.rs:33]: Yes, it does not make sense, sorry

File: src/db/users.rs (hunk 4 above)
Comment (luchoman08) [reply src/db/users.rs:9]: The only reason why I have not considered implementing `into_profile_for` in the `User` structure is because it would need a `Conn` reference as an argument, and I think a function that needs conn should not go in `models`. I hope to be corrected if I am wrong with that reasoning.

Comment (luchoman08) [reply src/db/users.rs:9]: I have summarized the implications of updating the type of the `author` property for `ArticleJson` and `CommentJson`, and the implications of the suggested function `into_profile_for` [in the next commit ](https://github.com/luchoman08/realworld-rust-rocket/commit/9fa0efe7f728f24522de31bcdbe30ca928632bbb) belonging to a temporary branch (includes the minor changes of the branch referred to in this pull request, and the recommendations made in previous comments) , if that branch is considered as the best candidate for pull request, I will cancel this pull request and move it to one made from the branch of the link.

Thanks and a thousand apologies for the inconvenience.

Comment (TatriX) [reply src/db/users.rs:9]: Makes sense. The only thing which bothers me is the function name.
Maybe something like `user_profile_for(conn, user, user_id)` would be more clear?

Comment (TatriX) [reply src/db/users.rs:9]: I checked another branch. It looks good to me and I'm ready to merge it.
If it's possible to split profile changes from clippy to a separate PR it would be really nice.

**Comment by TatriX** [comment]
Closing as stale.
